### PR TITLE
Flexible DKC

### PR DESCRIPTION
**Pull request by devrandom**
* Allows subclassing of DeterministicKeyChain, with factory for deserialization
* Allow overriding of derivation path

To consider:
* Should the derivation path be set in the builder, rather than through subclassing?

### CODE REVIEW

**Comment by mikehearn** [comment]
The above comments aren't exhaustive, let's think about the approach a bit first.

Is subclassing the right way to go here? It requires that we introduce a factory interface+class. Could we instead just have some kind of plugin object that does transformations on key paths? To what extent can the logic about what hierarchy layout to use encapsulateable? If we can extract it out into a separate class then it seems like it should be easier to use that API as you could just have a Bip44Hierarchy class or whatever, and it can be clearly documented in the javadocs what it's for.

**Comment by devrandom** [comment]
The approach you suggest sounds interesting.  I am would still want subclassing DKC because it's useful for other things.  Here is how we use it:

* Adjusting bloom filter handling (we use p2sh even for non-multisig)
* Vending of keys outside of KCG (we have two active keychains, one for assets and one for bitcoin)
* Derivation path

I look into the encapsulated hierarchy too.

**Comment by devrandom** [comment]
I rebased the stray commit out of existence, which made some of your line comments disappear since they were attached to specific commits...  sorry about that.  I did address all the line comments, including javadocs.

File: core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java [hunk 1]
```diff
@@ -1227,7 +1253,7 @@ public DeterministicSeed getSeed() {
         ImmutableList.Builder<DeterministicKey> keys = ImmutableList.builder();
         for (ECKey key : getKeys(true)) {
             DeterministicKey dKey = (DeterministicKey) key;
-            if (dKey.getPath().size() > 2) {
+            if (dKey.getPath().size() == getAccountPath().size() + 2) {
```
Comment (mikehearn) [inline core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java]: Hm is this change right? It's gone from everything below a certain path to just that single level.

File: core/src/test/java/org/bitcoinj/wallet/DeterministicKeyChainTest.java [hunk 2]
```diff
@@ -78,6 +79,86 @@ public void getKeys() throws Exception {
     }
 
     @Test
+    public void deriveAccountOne() throws Exception {
+        long secs = 1389353062L;
+        DeterministicKeyChain chain1 = new DeterministicKeyChain(ENTROPY, "", secs) {
```
Comment (mikehearn) [inline core/src/test/java/org/bitcoinj/wallet/DeterministicKeyChainTest.java]: You could use AccountOneChain here, no?

File: core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java (hunk 1 above)
Comment (devrandom) [inline core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java]: I think the old code was a bit too loose.  We know what level the leaves are, so referring to that specific level is better.

File: core/src/test/java/org/bitcoinj/wallet/DeterministicKeyChainTest.java (hunk 2 above)
Comment (devrandom) [inline core/src/test/java/org/bitcoinj/wallet/DeterministicKeyChainTest.java]: Agreed, forgot to refactor after I created the class.

**Comment by mikehearn** [comment]
This is merged, thanks! As it touches the key management code quite heavily, we will need some manual testing to go with the unit tests. Have you done any?

**Comment by devrandom** [comment]
Great.  I'll rebase the wallet that motivated this PR and do some manual
tests...

On 2015-06-03 03:52 AM, Mike Hearn wrote:
> This is merged, thanks! As it touches the key management code quite
> heavily, we will need some manual testing to go with the unit tests.
> Have you done any?
> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitcoinj/bitcoinj/pull/339#issuecomment-108296614>.
> 

-- 
devrandom / Miron

**Comment by devrandom** [comment]
I did a test with said wallet, and got a match after restoring from seed.  Probably a good next step is testing with Schildbach's.

**Comment by mikehearn** [comment]
OK, Andreas has pushed a version of his wallet that includes this code. I'll close this, thanks.

**Pull request closed without merging**